src/libsystemd-network/sd-dhcp-server.c

```diff
@@ -1397,3 +1482,71 @@ int sd_dhcp_server_set_relay_agent_information(
         free_and_replace(server->agent_remote_id, remote_id_dup);
         return 0;
 }
+
+int sd_dhcp_server_add_static_lease(
+                sd_dhcp_server *server,
+                const struct in_addr* address,
+                uint8_t *client_id,
+                size_t client_id_size) {
+         _cleanup_(dhcp_lease_freep) DHCPLease *lease = NULL, *old = NULL;
+        int r;
+
+        assert_return(server, -EINVAL);
+        assert_return(address, -EINVAL);
+        assert_return(client_id, -EINVAL);
+        assert_return(client_id_size == ETH_ALEN, -EINVAL);
+        assert_return(!sd_dhcp_server_is_running(server), -EBUSY);
+
+        lease = new(DHCPLease, 1);
+        if (!lease)
+                return -ENOMEM;
+
+        *lease = (DHCPLease) {
+                .address = address->s_addr,
+                .client_id.length = client_id_size,
+                .gateway = 0,
+                .expiration = 0,
+        };
+        lease->client_id.data = memdup(client_id, client_id_size);
+        if (!lease->client_id.data)
+                return -ENOMEM;
+
+        r = hashmap_ensure_allocated(&server->static_leases_by_client_id, &dhcp_lease_hash_ops);
+        if (r < 0)
+                return r;
+
+        old = hashmap_get(server->static_leases_by_client_id, &lease->client_id);
+        r = hashmap_replace(server->static_leases_by_client_id, &lease->client_id, lease);
```

Comment:
So, these three calls should be simply `hashmap_ensure_put()`. Then, if a user configure multiple static lease section in a .network file with the same MAC address, then the correct error will be propagated.